### DISCUSSION TITLE
failed validation prevents onChange from being called

### DISCUSSION THREAD

**Discussion by DashBarkHuss** [react-hook-form/react-hook-form#3863] [Q&A]
I'm trying to create an input that gets populated with data after the user scrapes a site for data: In this case a 3rd party shopping site for an item price. Then the user can edit the input.

<img width="286" alt="Screen Shot 2021-01-10 at 2 46 10 PM" src="https://user-images.githubusercontent.com/12969257/104133633-76c15880-534a-11eb-80b6-1410ddca4444.png">

To do this I make the input a controlled input. But now when the validation fails, `onChange` isn't called. `onChange` was where I called `setPrice`. So since `onChange` isn't called, the input can't be updated with what the user is typing.

<hr> 

## Steps to recreate:

 1. Go to the demo: [CodeSandbox](https://codesandbox.io/s/kind-bouman-9hrde?file=/src/App.js)
2. Click **Scrape site for item price**
 2. Click **Submit**
 3. Validation will fail, now try to add any character to the end of the input **19.99.** ex **19.99.1**.
 
Validation will fail again and the `onChange` handler won't be called. You will still see **19.99.**.

<hr> 

Even though validation fails, shouldn't the `onChange` handler still be called so the user can see what they are typing?

One way to solve this is to call `setPrice` in the validation when it fails or when there's an error `(errors.price || !isValidPrice(value)) setPrice(value);`. But this seems like a wonky solution. Is that the correct solution or is there a better way to solve this?

I know react-hook-form is a little weird with controlled inputs. But I thought that was only when you can't get access to the ref, but in this case I have access to the ref.

```javascript
import React, { useState } from "react";
import { useForm } from "react-hook-form";

const isValidPrice = (value) => {
  return (
    /^(0|[1-9][0-9]{0,2}(?:(.[0-9]{3})*|[0-9]*))(\,[0-9]+){0,1}$/.test(value) ||
    /^(0|[1-9][0-9]{0,2}(?:(,[0-9]{3})*|[0-9]*))(\.[0-9]+){0,1}$/.test(value)
  );
};

export default function App(props) {
  const [price, setPrice] = useState("");

  const { register, handleSubmit, errors } = useForm();

  const onSubmit = (data) => {
    console.log("submitted ", data);
  };
  const handleChange = (e) => {
    console.log("onchanged from ", price, e.target.value);
    setPrice(e.target.value);
    return true;
  };
  return (
    <>
      <button onClick={() => setPrice("19.99.")}>
        <span style={{ fontSize: ".7em" }}>(pretend to)</span> Scrape site for
        item price
      </button>
      <form autoComplete="off" onSubmit={handleSubmit(onSubmit)}>
        <p>Price 1</p>
        <input
          ref={register({
            validate: (value) => {
              console.log(
                "validation change from ",
                price,
                value,
                isValidPrice(value)
              );
              const valid = isValidPrice(value);

              // if (errors.price || !valid) setPrice(value);; //because onChange will not run
              return valid || `${value} is not a valid price.`;
            },
          })}
          name="price"
          id="price"
          value={price || ""}
          onChange={handleChange}
        />
        <br></br>
        Validation Errors: {errors.price?.message}
        <br></br>
        <button type="submit">Submit</button>
      </form>
    </>
  );
}
```

**Comment by bluebill1049** [comment]
i think you meant to use `setValue`. you can remove the onChange handler as well.

**Reply by bluebill1049** [reply]
```
<input
          ref={register({
            validate: (value) => {
              console.log(
                "validation change from ",
                price,
                value,
                isValidPrice(value)
              );
              const valid = isValidPrice(value);

              // if (errors.price || !valid) setPrice(value);; //because onChange will not run
              return valid || `${value} is not a valid price.`;
            }
          })}
          name="price"
          id="price"
        />
```

then
        
`setValue('price', 'xxx')` 

you can even trigger the validation as well `trigger('price')`
